core/drand_beacon.go

```diff
@@ -451,78 +455,100 @@ var errNoRoundInPeers = errors.New("could not find round")
 
 func (bp *BeaconProcess) storeCurrentFromPeerNetwork(ctx context.Context, store chain.Store) error {
 	clkNow := bp.opts.clock.Now().Unix()
-	currentRound := chain.CurrentRound(clkNow, bp.group.Period, bp.group.GenesisTime)
-	if currentRound < 1 {
+	if bp.group == nil {
+		return nil
+	}
+
+	targetRound := chain.CurrentRound(clkNow, bp.group.Period, bp.group.GenesisTime)
+	if targetRound < 1 {
 		// We cannot sync the initial round.
 		// Assume this is a fresh start
 		return nil
 	}
 
 	peers := bp.computePeers(bp.group.Nodes)
-	previousRound, err := bp.loadBeaconFromPeers(ctx, currentRound, peers)
+	targetBeacon, err := bp.loadBeaconFromPeers(ctx, targetRound, peers)
+	if errors.Is(err, errNoRoundInPeers) {
+		// If we can't find the desired beacon round, let's try with the previous one
+		if targetRound > 0 {
+			targetBeacon, err = bp.loadBeaconFromPeers(ctx, targetRound-1, peers)
+		}
+	}
+
 	if err != nil {
 		return err
 	}
 
 	verif := chain.NewVerifier(bp.group.Scheme)
-	err = verif.VerifyBeacon(previousRound, bp.group.PublicKey.Key())
+	err = verif.VerifyBeacon(targetBeacon, bp.group.PublicKey.Key())
 	if err != nil {
 		bp.log.Errorw("failed to verify beacon", "err", err)
 		return err
 	}
 
-	err = store.Put(ctx, &previousRound)
+	err = store.Put(ctx, &targetBeacon)
 	if err != nil {
 		bp.log.Errorw("failed to store beacon", "err", err)
 	}
 	return err
 }
 
 func (bp *BeaconProcess) loadBeaconFromPeers(ctx context.Context, targetRound uint64, peers []net.Peer) (chain.Beacon, error) {
-	found := false
-	round := chain.Beacon{}
+	select {
+	case <-ctx.Done():
+		return chain.Beacon{}, ctx.Err()
+	default:
+	}
+
+	type answer struct {
+		peer net.Peer
+		b    chain.Beacon
+		err  error
+	}
+
+	answers := make(chan answer, len(peers))
 
-	setMtx := sync.Mutex{}
+	// We should search for the beacon for three times the period of the network.
+	ctxFind, cancelFind := context.WithTimeout(ctx, bp.group.Period)
```

Comment:
Why not 3 * bp.group.Period then?

Reply:
That was supposed to be a TODO. What is a reasonable search period for the values? I'd say anything that's past 10s, as a hard limit, is too much. Is that a correct assumption? Should we use a more flexible timeout, like 3x group.Period? Or a similar value?

Reply:
I'd say 3 periods sounds reasonable, but so would 2.

Reply:
I guess if we go sub-seconds in our tests one day it might be a problem to have only relative times, so a fixed timeout is also fine. 1 period seems too little tho.